src/psyclone/psyir/nodes/operation.py

```diff
@@ -224,38 +242,209 @@ def create(operator, lhs, rhs):
+        `DeferredType` is returned.
+
+        :param argtypes: the types of the two operands.
+        :type argtypes: list[:py:class:`psyclone.psyir.symbols.DataType`,
+                             :py:class:`psyclone.psyir.symbols.DataType`]
+
+        :returns: the base type of the result of the operation.
+        :rtype: :py:class:`psyclone.psyir.symbols.DataType`
+
+        :raises TypeError: if an unexpected intrinsic type is found for
+                           either of the operands to a numeric operation.
+
+        '''
+        if self.operator not in self._numeric_ops:
+            # Must be a relational or logical operator. Intrinsic type of
+            # result will be boolean.
+            return BOOLEAN_TYPE
+
+        # We have a numerical operation.
+        if any(isinstance(atype.intrinsic, DeferredType)
+               for atype in argtypes):
+            # datatype of a numerical operation on a DeferredType is a
+            # DeferredType.
+            return DeferredType()
+
+        base_type = None
+
+        # If either of the operands has REAL intrinsic type then the result
+        # must also be REAL.
+        if argtypes[0].intrinsic == argtypes[1].intrinsic:
+            # Operands are of the same intrinsic type.
+            precision = self._get_result_precision([argtypes[0].precision,
+                                                    argtypes[1].precision])
+            base_type = ScalarType(argtypes[0].intrinsic, precision)
+        elif argtypes[0].intrinsic == ScalarType.Intrinsic.REAL:
+            base_type = argtypes[0]
+        elif argtypes[1].intrinsic == ScalarType.Intrinsic.REAL:
+            base_type = argtypes[1]
+
+        # Check that the type of the result is consistent with a numerical
+        # operation.
+        if not base_type or base_type.intrinsic not in (
+                ScalarType.Intrinsic.INTEGER,
+                ScalarType.Intrinsic.REAL):
+            for atype in argtypes:
+                if atype.intrinsic not in (ScalarType.Intrinsic.INTEGER,
+                                           ScalarType.Intrinsic.REAL):
+                    raise TypeError(
+                        f"Invalid argument of type '{atype.intrinsic}' to "
+                        f"numerical operation '{self.operator}' in "
+                        f"'{self.debug_string()}'. Currently only "
+                        f"ScalarType.Intrinsic.REAL/INTEGER are "
+                        f"supported (TODO #1590)")
+        return base_type
+
+    @property
+    def datatype(self):
+        '''
+        Determines the datatype of this operation. If it cannot be determined
+        for any reason then an instance of DeferredType is returned.
+
+        :returns: the datatype of the result of this BinaryOperation.
+        :rtype: :py:class:`psyclone.psyir.symbols.DataType`
+
+        :raises TypeError: if the operands are both arrays but are of different
+                           shapes.
```

Comment:
I agree with the previous response that I resolved about this being an error an not a DeferredTpe, but I believe this should be a InternalError.

Reply:
Yes, done.